File: exercise/45/solution.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int n = nums.size();
        if (n <= 1) return 0;
        int current = 0, pre = 0, jump = 0;
        for (int i = 0; i < n; i++) {
            if (i > pre) {
                jump++;
                pre = current;
            }
            if (i + nums[i] >= n - 1)
                return jump + 1;
            if (i + nums[i] > current) {
                current = i + nums[i];
            }
        }
        return -1;
    }
};

File: exercise/45/solution.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return 0
        edge = 0
        steps = [0 for _ in range(len(nums))] 
        for i in range(len(nums)):
            if i + nums[i] >= len(nums) - 1:
                return steps[i] + 1
            elif i + nums[i] > edge:
                for j in range(edge + 1, i + nums[i] + 1):
                    steps[j] = steps[i] + 1
                edge = i + nums[i]
        return steps[-1]
